src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,330 @@
+        /// <param name="numberOfOccurrences">The number of occurrences between the time and the recurrence start.</param>
+        /// </summary>
+        private static void FindDailyPreviousOccurrence(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset previousOccurrence, out int numberOfOccurrences)
+        {
+            RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+            DateTimeOffset start = settings.Start.Value;
+
+            Debug.Assert(time >= start);
+
+            int interval = pattern.Interval;
+
+            TimeSpan timeGap = time - start;
+
+            //
+            // netstandard2.0 does not support '/' operator for TimeSpan. After we stop supporting netstandard2.0, we can remove .TotalSeconds.
+            int numberOfInterval = (int)Math.Floor(timeGap.TotalSeconds / TimeSpan.FromDays(interval).TotalSeconds);
+
+            previousOccurrence = start.AddDays(numberOfInterval * interval);
+
+            numberOfOccurrences = numberOfInterval + 1;
+        }
+
+        /// <summary>
+        /// Find the closest previous recurrence occurrence before the provided timestamp according to the "Weekly" recurrence pattern.
+        /// <param name="time">A timestamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <param name="previousOccurrence">The closest previous occurrence.</param>
+        /// <param name="numberOfOccurrences">The number of occurrences between the time and the recurrence start.</param>
+        /// </summary>
+        private static void FindWeeklyPreviousOccurrence(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset previousOccurrence, out int numberOfOccurrences)
+        {
+            RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+            DateTimeOffset start = settings.Start.Value;
+
+            Debug.Assert(time >= start);
+
+            int interval = pattern.Interval;
+
+            DateTimeOffset firstDayOfStartWeek = start.AddDays(
+                -CalculateWeeklyDayOffset(start.DayOfWeek, pattern.FirstDayOfWeek));
+
+            //
+            // netstandard2.0 does not support '/' operator for TimeSpan. After we stop supporting netstandard2.0, we can remove .TotalSeconds.
+            int numberOfInterval = (int)Math.Floor((time - firstDayOfStartWeek).TotalSeconds / TimeSpan.FromDays(interval * DaysPerWeek).TotalSeconds);
+
+            DateTimeOffset firstDayOfMostRecentOccurringWeek = firstDayOfStartWeek.AddDays(numberOfInterval * (interval * DaysPerWeek));
+
+            List<DayOfWeek> sortedDaysOfWeek = SortDaysOfWeek(pattern.DaysOfWeek, pattern.FirstDayOfWeek);
+
+            //
+            // Subtract the days before the start in the first week.
+            numberOfOccurrences = numberOfInterval * sortedDaysOfWeek.Count - sortedDaysOfWeek.IndexOf(start.DayOfWeek);
```

Comment:
For weekly, does number of occurences represent how many weeks pass? Or how many individual windows pass?

For example, if I have Weekly with Mon, Wed, Fri and number of occurrences is two, does that mean I'll have two separate weeks with Mon, Wed, Fri or a single week with only two days?

Reply:
Number of occurrences represent how many individual windows **start time** have passed.

> For example, if I have Weekly with Mon, Wed, Fri and number of occurrences is two, does that mean I'll have two separate weeks with Mon, Wed, Fri or a single week with only two days?

a single week with only two days 

It could also be other case, for example: today is 2024.3.20 (Wed), 
Weekly Pattern is [Mon, Tue, Thu] and Interval is 2
The start is 2024.3.7 (Thu).

For today (2024.3.20 (Wed)), the number of occurrences is 3 (but it actualy went across 3 weeks)